core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -893,50 +893,87 @@ class KafkaApis(val requestChannel: RequestChannel,
 
     val offsetFetchResponse =
-    // reject the request if not authorized to the group
-    if (!authorize(request.session, Read, new Resource(Group, offsetFetchRequest.groupId))) {
-      val unauthorizedGroupResponse = new OffsetFetchResponse.PartitionData(OffsetFetchResponse.INVALID_OFFSET, "", Errors.GROUP_AUTHORIZATION_FAILED.code)
-      val results = offsetFetchRequest.partitions.asScala.map { topicPartition => (topicPartition, unauthorizedGroupResponse)}.toMap
-      new OffsetFetchResponse(results.asJava)
-    } else {
-      val (authorizedTopicPartitions, unauthorizedTopicPartitions) = offsetFetchRequest.partitions.asScala.partition { topicPartition =>
-        authorize(request.session, Describe, new Resource(auth.Topic, topicPartition.topic))
-      }
-      val unknownTopicPartitionResponse = new OffsetFetchResponse.PartitionData(OffsetFetchResponse.INVALID_OFFSET, "", Errors.UNKNOWN_TOPIC_OR_PARTITION.code)
-      val unauthorizedStatus = unauthorizedTopicPartitions.map(topicPartition => (topicPartition, unknownTopicPartitionResponse)).toMap
-
-      if (header.apiVersion == 0) {
-        // version 0 reads offsets from ZK
-        val responseInfo = authorizedTopicPartitions.map { topicPartition =>
-          val topicDirs = new ZKGroupTopicDirs(offsetFetchRequest.groupId, topicPartition.topic)
-          try {
-            if (!metadataCache.contains(topicPartition.topic))
-              (topicPartition, unknownTopicPartitionResponse)
-            else {
-              val payloadOpt = zkUtils.readDataMaybeNull(s"${topicDirs.consumerOffsetDir}/${topicPartition.partition}")._1
-              payloadOpt match {
-                case Some(payload) =>
-                  (topicPartition, new OffsetFetchResponse.PartitionData(payload.toLong, "", Errors.NONE.code))
-                case None =>
-                  (topicPartition, unknownTopicPartitionResponse)
+      // reject the request if not authorized to the group
+      if (!authorize(request.session, Read, new Resource(Group, offsetFetchRequest.groupId))) {
+        if (header.apiVersion >= 2)
+          new OffsetFetchResponse(Errors.GROUP_AUTHORIZATION_FAILED)
+        else {
+          val unauthorizedGroupResponse = new OffsetFetchResponse.PartitionData(
+              OffsetFetchResponse.INVALID_OFFSET, OffsetFetchResponse.NO_METADATA, Errors.GROUP_AUTHORIZATION_FAILED)
+          val results = offsetFetchRequest.partitions.asScala.map { topicPartition => (topicPartition, unauthorizedGroupResponse)}.toMap
+          new OffsetFetchResponse(results.asJava, Errors.NONE, header.apiVersion)
+        }
+      } else {
+        val isAllPartitions = offsetFetchRequest.isAllPartitions() // header.apiVersion == 2
+        var groupOffsets: (Errors, Map[TopicPartition, OffsetFetchResponse.PartitionData]) = null
+        val partitions =
+          if (isAllPartitions) {
+            groupOffsets = coordinator.handleFetchOffsets(offsetFetchRequest.groupId, null)
```

Comment:
Seems like this and the other call to `handleFetchOffsets` below needs to go to the `else` case after the check for version 0. For version 0, we pull offsets from zookeeper. I'm wondering if your first approach, which collected all partitions from the coordinator first, may have been a little cleaner. Another option to consider, perhaps you could do the post-filtering for the `isAllPartitions` case separately, and continue doing pre-filtering when we are provided the partition list.